lib/json_serialization.nit

```diff
@@ -219,47 +219,81 @@ redef class NativeString
 	redef fun serialize_to_json(v) do to_s.serialize_to_json(v)
 end
 
-redef class Array[E]
-	redef fun serialize_to_json(v)
+redef class Collection[E]
+	# Utility to serialize a normal Json array
+	private fun serialize_to_pure_json(v: JsonSerializer)
 	do
-		if class_name == "Array[nullable Serializable]" then
-			# Using class_name to the the exact type
-			# We do not want Array[Int] or anything else here
 			v.stream.write "["
 			var is_first = true
 			for e in self do
 				if is_first then
 					is_first = false
-				else v.stream.write(", ")
+				else v.stream.write ", "
 
 				if not v.try_to_serialize(e) then
 					v.warn("element of type {e.class_name} is not serializable.")
 				end
 			end
 			v.stream.write "]"
+	end
+end
+
+redef class Array[E]
+	redef fun serialize_to_json(v)
+	do
+		if class_name == "Array[nullable Serializable]" then
+			# Using class_name to get the exact type,
+			# we do not want Array[Int] or anything else here.
+
+			serialize_to_pure_json v
 		else
 			# Register as pseudo object
 			var id = v.ref_id_for(self)
 			v.stream.write "\{\"__kind\": \"obj\", \"__id\": {id}, \"__class\": \"{class_name}\""
-			v.stream.write """, "__length": {{{length}}}, "__items": ["""
-			var is_first = true
-			for e in self do
-				if is_first then
-					is_first = false
-				else v.stream.write(", ")
+			v.stream.write """, "__length": {{{length}}}, "__items": """
+			serialize_to_pure_json v
+			v.stream.write "\}"
+		end
+	end
 
-				if not v.try_to_serialize(e) then
-					v.warn("element of type {e.class_name} is not serializable.")
-				end
+	# Instantiate a new `Array` from its serialized representation.
+	redef init from_deserializer(v: Deserializer)
+	do
+		if v isa JsonDeserializer then
+			v.notify_of_creation self
+
+			var length = v.deserialize_attribute("__length").as(Int)
+			var arr = v.path.last["__items"].as(SequenceRead[nullable Object])
+			for i in length.times do
+				var obj = v.convert_object(arr[i])
+				self.add obj
 			end
-			v.stream.write "]"
-			v.stream.write "\}"
 		end
 	end
+end
+
+redef class Set[E]
+	redef fun serialize_to_json(v)
+	do
+		# Register as pseudo object
+		var id = v.ref_id_for(self)
+		v.stream.write """{"__kind": "obj", "__id": """
+		v.stream.write id.to_s
+		v.stream.write """, "__class": """"
```

Comment:
Yup, I still don't know how I'll be able to fix it with the colouring syntax of github, unfortunately.

Though, for simple cases like this, I think we'd be better off simply using single-quoted strings and escaping the few uses of `"` in it